frontend/app_flowy/packages/flowy_editor/lib/document/text_delta.dart

```diff
@@ -0,0 +1,399 @@
+import 'dart:math';
+
+import 'package:flutter/foundation.dart';
+import './node.dart';
+
+class TextOperation {
+  bool get isEmpty {
+    return length == 0;
+  }
+
+  int get length {
+    return 0;
+  }
+
+  Attributes? get attributes {
+    return null;
+  }
+}
+
+int _hashAttributes(Attributes attributes) {
+  return Object.hashAllUnordered(attributes.entries);
+}
+
+class TextInsert extends TextOperation {
+  String content;
+  final Attributes? _attributes;
+
+  TextInsert(this.content, [Attributes? attrs]) : _attributes = attrs;
+
+  @override
+  int get length {
+    return content.length;
+  }
+
+  @override
+  Attributes? get attributes {
+    return _attributes;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! TextInsert) {
+      return false;
+    }
+    return content == other.content &&
+        mapEquals(_attributes, other._attributes);
+  }
+
+  @override
+  int get hashCode {
+    final contentHash = content.hashCode;
+    final attrs = _attributes;
+    return Object.hash(
+        contentHash, attrs == null ? null : _hashAttributes(attrs));
+  }
+}
+
+class TextRetain extends TextOperation {
+  int _length;
+  final Attributes? _attributes;
+
+  TextRetain({
+    required length,
+    attributes,
+  })  : _length = length,
+        _attributes = attributes;
+
+  @override
+  bool get isEmpty {
+    return length == 0;
+  }
+
+  @override
+  int get length {
+    return _length;
+  }
+
+  set length(int v) {
+    _length = v;
+  }
+
+  @override
+  Attributes? get attributes {
+    return _attributes;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! TextRetain) {
+      return false;
+    }
+    return _length == other.length && mapEquals(_attributes, other._attributes);
+  }
+
+  @override
+  int get hashCode {
+    final attrs = _attributes;
+    return Object.hash(_length, attrs == null ? null : _hashAttributes(attrs));
+  }
+}
+
+class TextDelete extends TextOperation {
+  int _length;
+
+  TextDelete({
+    required int length,
+  }) : _length = length;
+
+  @override
+  bool get isEmpty {
+    return length == 0;
+  }
+
+  @override
+  int get length {
+    return _length;
+  }
+
+  set length(int v) {
+    _length = v;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! TextDelete) {
+      return false;
+    }
+    return _length == other.length;
+  }
+
+  @override
+  int get hashCode {
+    return _length.hashCode;
+  }
+}
+
+class _OpIterator {
+  final List<TextOperation> _operations;
+  int _index = 0;
+  int _offset = 0;
+
+  _OpIterator(List<TextOperation> operations) : _operations = operations;
```

Comment:
It seems like the _OpIterator will change the passed in value's length. Is this ok?
if yes, the iterator should not modify the passed in vec.